man/seaudit.8
.TH seaudit 8
.SH NAME
seaudit \- SELinux graphical audit log analysis tool
.SH SYNOPSIS
.B seaudit
[OPTIONS] [POLICY ...]
.SH DESCRIPTION
.PP
.B seaudit
allows the user to view and filter the contents of a log file.
.B seaudit
supports the syslog and auditd log formats and provides queries to inspect the SELinux policy based on log messages.
.SH POLICY
.PP
.B
seaudit
supports loading a SELinux policy in one of four formats.
.IP "source"
A single text file containing policy source for versions 12 through 21. This file is usually named policy.conf.
.IP "binary"
A single file containing a monolithic kernel binary policy for versions 15 through 21. This file is usually named by version - for example, policy.20.
.IP "modular"
A list of policy packages each containing a loadable policy module. The first module listed must be a base module.
.IP "policy list"
A single text file containing all the information needed to load a policy, usually exported by SETools graphical utilities.
.PP
If no policy file is provided,
.B
seaudit
will search for the system default policy: checking first for a source policy, next for a binary policy matching the running kernel's preferred version, and finally for the highest version that can be found.
If no policy can be found,
.B
seaudit
will begin with no policy loaded.
.SH OPTIONS
.IP "-l FILE, --log=FILE"
Upon startup, open the log FILE instead of the system log file.
.IP "-h, --help"
Print help information and exit.
.IP "-V, --version"
Print version information and exit.
.SH AUTHOR
This manual page was written by Jeremy A. Mowery <jmowery@tresys.com>.
.SH COPYRIGHT
Copyright(C) 2006-2007 Tresys Technology, LLC
.SH BUGS
Please report bugs via an email to setools-bugs@tresys.com.
.SH SEE ALSO
seaudit-report(8)


NEWS
SETools 3.2:

libapol, libqpol, libesaudit, libpoldiff:

  * If --enable-swig-python is given during configure time, the build
    system will create Python SWIG wrappers for these libraries.

  * If --enable-swig-java is given during configure time, the build
    system will create Java SWIG wrappers for these libraries.

libpoldiff:

  * Provides ability to diff levels, categories, and range_transition
    statements.  Provides ability to diff modifications to a user's
    default level and permitted MLS range, assuming the policies are MLS.
    
apol, seaudit, sediffx:

  * Introduces 'policy list', a small text file that contains
    references to a base policy and any number of modules.  After
    selecting the base policy and modules in a tool's open policy
    dialog, click on 'Export' to write a policy list to disk.  That
    policy list then may be imported into the same tool or any other
    graphical SETools application.  This file also may be specified
    on the command line for all tools that load a policy.

sesearch, apol:

  * Provides full support for version 21 policy (i.e., object classes for
    range_transition statements).

sediff:

  * Shows MLS diffs.

sediffx:

  * Shows MLS diffs.

  * Provides option to show line numbers that contributed just to a
    specific AV rule's permission by clicking that AV rule's permissions.
    This is in addition to existing sediffx behavior that showed all lines
    associated with a particular rule difference.

  * Provides support for type joins and splits within the type remap dialog.

==================================================

SETools 3.1:

SETools:

  * All tools that open a policy now support loadable policy modules.
    Command line tools expect the first module to be a base module
    followed optionally by any other modules.  Graphical tools have
    a new open policy dialog to select a base module and any number of
    additional modules.

  * Release of RPM packages that are compatible with Fedora Core 5 and
    6.  The spec and support files are in packages/rpm.

libapol:

  * New class apol_policy_path_t to represent a base policy and any
    number of modules.  Use this whenever referring to the file or
    files constituting a policy.
  
libqpol:

  * Policy features such as attribute names or MLS can now be queried
    individally via qpol_policy_has_capability() rather than inferred
    by policy type and version.

  * New class qpol_module_t to represent a particular policy module
    prior to it being linked into a base policy (qpol_policy_t).

libseaudit:

  * Rewrite of library to have proper namespaces.  libseaudit is now
    fully documented and suitable for third-party users.

seaudit:

  * Rewrite to use new libseaudit.

  * Numerous tweaks to the interface to be more user friendly.

seaudit-report:

  * Rewrite to use new libseaudit.

sediffx:

  * Numerous tweaks to the interface to be more user friendly.

==================================================

SETools 3.0.1

SETools:

  * All code has been indented uniformly via the 'make indent' target.

SETools libraries:

  * All libraries now have a get_version() function.

libqpol:

  * Syntactic rule table is now off by default; it requires an
    explicit call to qpol_policy_build_syn_rule_table() to create it.

libapol:

  * apol_policy_t is now an opaque structure.  apol_permmap_t is no
    longer a public declaration.

  * avrule and terule queries now have full "syntactic" searching
    features.

apol & sesearch:

  * If loaded policy is source, the new syntactic search algorithm is used.

==================================================

SETools 3.0

SETools:

  * Moved entire project to autoconf build system.  This will detect
    dependencies correctly and make it easier to integrate with Linux
    distributions and their packaging systems.

SETools libraries:

  * Rewrite of back-end of SETools to use libsepol data structures.
    Most things should run notably faster.

  * All exported library functions have a standardized naming
    convention and are fully documented.  This will prove helpful for
    third-party developers integrating SETools into their own
    projects.

  * For policies version 15 or greater, domain transition analysis now
    takes into consideration setexec permission and/or type_transition
    rules.

apol:

  * Rework Tk interface to fit on 1024x768 displays.

  * Indirect matching of attributes now works with rule searches.

sediff:

  * New diffing algorithm yields significant speed improvements.

  * Can now diff neverallow and role_transition statements.

  * Streamlined results display - should be easier to read.

sechecker:

  * Updated module format and template for ease of extension.


=======================================================
May 1, 2006, Version 2.4

apol:
	File contexts tab now allows for MLS range searching if 
	the loaded database is from a MLS filesystem.
	Policy statistics dialog now shows MLS and ocontexts
	summaries.

libapol:
	Added support for loading base policies containing optionals.
	Added support for searching range transitions containing
	attributes.

libseaudit:
	Bugfix to support parsing FC5-style audit logs.

seaudit:
	Added date filters.

secmds:
	Added support to indexcon and searchcon for MLS filesytems.
	Added support to findcon and replcon for MLS filesystems.
	
sechecker:
	Added incomplete network access (inc_net_access) module.
	Added unreachable domains (unreachable_doms) module.
	Added impossible range transitions (imp_range_trans) module.

sesearch:
	Allow user to search range transitions by attributes and
	indirect matching.
	Added RBAC searching.


=======================================================
January 23, 2006, Version 2.3

apol:
	added new MLS components tab for sensitivities, 
	levels, and categories.
	changed users tab to support ranges and default 
	levels.
	added range transition tab for searching range
	transition rules.

seaudit/seaudit.c
/**
 *  @file
 *  Main driver for the seaudit application.  This file also
 *  implements the main class seaudit_t.
 *
 *  @author Jeremy A. Mowery jmowery@tresys.com
 *  @author Jason Tang jtang@tresys.com
 *
 *  Copyright (C) 2003-2007 Tresys Technology, LLC
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>

#include "seaudit.h"
#include "toplevel.h"

#include <apol/util.h>
#include <seaudit/model.h>
#include <seaudit/parse.h>
#include <seaudit/util.h>

#include <errno.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <glade/glade.h>
#include <glib.h>
#include <gtk/gtk.h>

struct seaudit
{
	preferences_t *prefs;
	apol_policy_t *policy;
	apol_policy_path_t *policy_path;
	seaudit_log_t *log;
	FILE *file;
	char *log_path;
	size_t num_log_messages;
	struct tm *first, *last;
	toplevel_t *top;
};

static struct option const opts[] = {
	{"log", required_argument, NULL, 'l'},
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'V'},
	{NULL, 0, NULL, 0}
};

preferences_t *seaudit_get_prefs(seaudit_t * s)
{
	return s->prefs;
}

void seaudit_set_policy(seaudit_t * s, apol_policy_t * policy, apol_policy_path_t * path)
{
	if (policy != NULL) {
		if (preferences_add_recent_policy(s->prefs, path) < 0) {
			toplevel_ERR(s->top, "%s", strerror(errno));
			apol_policy_destroy(&policy);
			return;
		}
		apol_policy_destroy(&s->policy);
		s->policy = policy;
		if (path != s->policy_path) {
			apol_policy_path_destroy(&s->policy_path);
		}
		s->policy_path = path;
	} else {
		apol_policy_destroy(&s->policy);
		apol_policy_path_destroy(&s->policy_path);
	}
}

apol_policy_t *seaudit_get_policy(seaudit_t * s)
{
	return s->policy;
}

apol_policy_path_t *seaudit_get_policy_path(seaudit_t * s)
{
	return s->policy_path;
}

void seaudit_set_log(seaudit_t * s, seaudit_log_t * log, FILE * f, const char *filename)
{
	if (s->file != NULL) {
		fclose(s->file);
		s->file = NULL;
	}
	if (log != NULL) {
		seaudit_model_t *model = NULL;
		apol_vector_t *messages = NULL;
		char *t = NULL;
		if ((model = seaudit_model_create(NULL, log)) == NULL ||
		    (messages = seaudit_model_get_messages(log, model)) == NULL ||
		    (t = strdup(filename)) == NULL || preferences_add_recent_log(s->prefs, filename) < 0) {
			toplevel_ERR(s->top, "%s", strerror(errno));
			seaudit_log_destroy(&log);
			seaudit_model_destroy(&model);
			apol_vector_destroy(&messages);
			free(t);
			return;
		}
		/* do it in this order, for filename could be pointing to
		 * s->log_path */
		seaudit_log_destroy(&s->log);
		s->log = log;
		s->file = f;
		free(s->log_path);
		s->log_path = t;
		s->num_log_messages = apol_vector_get_size(messages);
		if (s->num_log_messages == 0) {
			s->first = s->last = NULL;
		} else {
			seaudit_message_t *message = apol_vector_get_element(messages, 0);
			s->first = seaudit_message_get_time(message);
			message = apol_vector_get_element(messages, s->num_log_messages - 1);
			s->last = seaudit_message_get_time(message);
		}
		seaudit_model_destroy(&model);
		apol_vector_destroy(&messages);
	} else {
		seaudit_log_destroy(&s->log);
		free(s->log_path);
		s->log_path = NULL;
		s->num_log_messages = 0;
		s->first = s->last = NULL;
	}
}

int seaudit_parse_log(seaudit_t * s)
{
	return seaudit_log_parse(s->log, s->file);
}

seaudit_log_t *seaudit_get_log(seaudit_t * s)
{
	return s->log;
}

char *seaudit_get_log_path(seaudit_t * s)
{
	return s->log_path;
}

apol_vector_t *seaudit_get_log_users(seaudit_t * s)
{
	if (s->log == NULL) {
		return NULL;
	} else {
		return seaudit_log_get_users(s->log);
	}
}

apol_vector_t *seaudit_get_log_roles(seaudit_t * s)
{
	if (s->log == NULL) {
		return NULL;
	} else {
		return seaudit_log_get_roles(s->log);
	}
}

apol_vector_t *seaudit_get_log_types(seaudit_t * s)
{
	if (s->log == NULL) {
		return NULL;
	} else {
		return seaudit_log_get_types(s->log);
	}
}

apol_vector_t *seaudit_get_log_classes(seaudit_t * s)
{
	if (s->log == NULL) {
		return NULL;
	} else {
		return seaudit_log_get_classes(s->log);
	}
}

size_t seaudit_get_num_log_messages(seaudit_t * s)
{
	return s->num_log_messages;
}

struct tm *seaudit_get_log_first(seaudit_t * s)
{
	return s->first;
}

struct tm *seaudit_get_log_last(seaudit_t * s)
{
	return s->last;
}

static seaudit_t *seaudit_create(preferences_t * prefs)
{
	seaudit_t *s = calloc(1, sizeof(*s));
	if (s != NULL) {
		s->prefs = prefs;
	}
	return s;
}

static void seaudit_destroy(seaudit_t ** s)
{
	if (s != NULL && *s != NULL) {
		apol_policy_destroy(&(*s)->policy);
		seaudit_log_destroy(&(*s)->log);
		if ((*s)->file != NULL) {
			fclose((*s)->file);
		}
		preferences_destroy(&(*s)->prefs);
		toplevel_destroy(&(*s)->top);
		free((*s)->policy_path);
		free((*s)->log_path);
		free(*s);
		*s = NULL;
	}
}

static void print_version_info(void)
{
	printf("seaudit %s\n%s\n", VERSION, COPYRIGHT_INFO);
}

static void print_usage_info(const char *program_name, int brief)
{
	printf("Usage: %s [OPTIONS] [POLICY ...]\n\n", program_name);
	if (brief) {
		printf("\tTry %s --help for more help.\n\n", program_name);
		return;
	}
	printf("Audit Log analysis tool for Security Enhanced Linux.\n\n");
	printf("   -l FILE, --log=FILE     open the log FILE\n");
	printf("   -h, --help              print this help text and exit\n");
	printf("   -V, --version           print version information and exit\n\n");
}

static void seaudit_parse_command_line(seaudit_t * seaudit, int argc, char **argv, const char **log, apol_policy_path_t ** policy)
{
	int optc;
	*log = NULL;
	*policy = NULL;
	apol_policy_path_type_e path_type = APOL_POLICY_PATH_TYPE_MONOLITHIC;
	char *primary_path = NULL;
	apol_vector_t *modules = NULL;
	while ((optc = getopt_long(argc, argv, "l:hV", opts, NULL)) != -1) {
		switch (optc) {
		case 'l':{
				*log = optarg;
				break;
			}
		case 'h':{
				print_usage_info(argv[0], 0);
				seaudit_destroy(&seaudit);
				exit(EXIT_SUCCESS);
			}
		case 'V':{
				print_version_info();
				seaudit_destroy(&seaudit);
				exit(EXIT_SUCCESS);
			}
		default:{
				/* unrecognized argument give full usage */
				print_usage_info(argv[0], 1);
				seaudit_destroy(&seaudit);
				exit(EXIT_FAILURE);
			}
		}
	}
	if (optind < argc) {	       /* modules */
		if ((modules = apol_vector_create(NULL)) == NULL) {
			ERR(NULL, "%s", strerror(ENOMEM));
			seaudit_destroy(&seaudit);
			exit(EXIT_FAILURE);
		}
		path_type = APOL_POLICY_PATH_TYPE_MONOLITHIC;
		primary_path = argv[optind++];
		while (argc - optind) {
			if (apol_vector_append(modules, argv[optind])) {
				ERR(NULL, "%s", strerror(ENOMEM));
				seaudit_destroy(&seaudit);
				exit(EXIT_FAILURE);
			}
			path_type = APOL_POLICY_PATH_TYPE_MODULAR;
			optind++;
		}
	}
	if (*log == NULL) {
		*log = preferences_get_log(seaudit->prefs);
	}
	if (primary_path != NULL && strcmp(primary_path, "") != 0) {
		if (apol_file_is_policy_path_list(primary_path)) {
			if ((*policy = apol_policy_path_create_from_file(primary_path)) == NULL) {
				ERR(NULL, "%s", "invalid policy list");
				seaudit_destroy(&seaudit);
				exit(EXIT_FAILURE);
			}
		} else {
			if ((*policy = apol_policy_path_create(path_type, primary_path, modules)) == NULL) {
				ERR(NULL, "%s", strerror(ENOMEM));
				seaudit_destroy(&seaudit);
				exit(EXIT_FAILURE);
			}
		}
	} else {
		const apol_policy_path_t *path = preferences_get_policy(seaudit->prefs);
		if (path != NULL && (*policy = apol_policy_path_create_from_policy_path(path)) == NULL) {
			ERR(NULL, "%s", strerror(ENOMEM));
			seaudit_destroy(&seaudit);
			exit(EXIT_FAILURE);
		}
	}
	apol_vector_destroy(&modules);
}

/*
 * We don't want to do the heavy work of loading and displaying the
 * log and policy before the main loop has started because it will
 * freeze the gui for too long.  To solve this, the function is called
 * from an idle callback set-up in main.
 */
struct delay_file_data
{
	toplevel_t *top;
	const char *log_filename;
	apol_policy_path_t *policy_path;
};

static gboolean delayed_main(gpointer data)
{
	struct delay_file_data *dfd = (struct delay_file_data *)data;
	if (dfd->log_filename != NULL && strcmp(dfd->log_filename, "") != 0) {
		toplevel_open_log(dfd->top, dfd->log_filename);
	}
	if (dfd->policy_path != NULL) {
		toplevel_open_policy(dfd->top, dfd->policy_path);
	}
	return FALSE;
}

int main(int argc, char **argv)
{
	preferences_t *prefs;
	seaudit_t *app;
	const char *log;
	apol_policy_path_t *policy;
	struct delay_file_data file_data;

	gtk_init(&argc, &argv);
	glade_init();
	if (!g_thread_supported())
		g_thread_init(NULL);
	if ((prefs = preferences_create()) == NULL) {
		ERR(NULL, "%s", strerror(ENOMEM));
		exit(EXIT_FAILURE);
	}
	if ((app = seaudit_create(prefs)) == NULL) {
		ERR(NULL, "%s", strerror(ENOMEM));
		exit(EXIT_FAILURE);
	}
	seaudit_parse_command_line(app, argc, argv, &log, &policy);
	if ((app->top = toplevel_create(app)) == NULL) {
		ERR(NULL, "%s", strerror(ENOMEM));
		seaudit_destroy(&app);
		exit(EXIT_FAILURE);
	}
	file_data.top = app->top;
	file_data.log_filename = log;
	file_data.policy_path = policy;
	g_idle_add(&delayed_main, &file_data);
	gtk_main();
	if (preferences_write_to_conf_file(app->prefs) < 0) {
		ERR(NULL, "%s", strerror(ENOMEM));
	}
	seaudit_destroy(&app);
	exit(EXIT_SUCCESS);
}
